src/mito2/src/compaction.rs

```diff
@@ -495,23 +513,36 @@ impl CompactionStatus {
             region_id,
             version_control,
             access_layer,
-            pending_compaction: None,
+            waiters: Vec::new(),
+            pending_request: None,
         }
     }
 
-    /// Merge the watier to the pending compaction.
-    fn merge_waiter(&mut self, waiter: OptionOutputTx) {
-        let pending = self
-            .pending_compaction
-            .get_or_insert_with(|| PendingCompaction {
-                waiters: Vec::new(),
-            });
-        pending.push_waiter(waiter);
+    /// Merge the waiter to the pending compaction.
+    fn merge_waiter(&mut self, mut waiter: OptionOutputTx) {
+        if let Some(waiter) = waiter.take_inner() {
+            self.waiters.push(waiter);
+        }
+    }
+
+    /// Set pending compaction request or replace current value if already exist.
+    fn set_pending_request(&mut self, pending: PendingCompaction) {
+        if let Some(mut prev) = self.pending_request.replace(pending) {
+            debug!(
+                "Replace pending compaction options with new request {:?} for region: {}",
+                prev.options, self.region_id
+            );
+            if let Some(waiter) = prev.waiter.take_inner() {
+                waiter.send(ManualCompactionOverrideSnafu.fail());
+            }
+        }
     }
 
-    fn on_failure(self, err: Arc<Error>) {
-        if let Some(mut pending) = self.pending_compaction {
-            pending.on_failure(self.region_id, err.clone());
+    fn on_failure(mut self, err: Arc<Error>) {
+        for waiter in self.waiters.drain(..) {
+            waiter.send(Err(err.clone()).context(CompactRegionSnafu {
+                region_id: self.region_id,
+            }));
```

Comment:
Do we need to also cancel the pending compaction?

Reply:
> Do we need to also cancel the pending compaction?

fixed in 6b71d0b911